### ISSUE TITLE
joni interprets `[\w-#]` as `[ !"#0-9A-Z_a-z]`

### ISSUE THREAD

**Issue by haozhun** [jruby/joni#14]
Joni interprets `[\w-#]` as `[ !"#0-9A-Z_a-z]` in both default syntax and Java syntax. Java Pattern interprets it as `[-0-9A-Z_a-z]`.

An addition question: in general, is it considered a bug if interpretation doesn't match Java pattern when syntax is set to Java?

**Comment by headius** [comment]
Yes, if the Java dialect does not match Java pattern behavior, we should fix that.
